### PR TITLE
Scrum 18  api post send email

### PR DESCRIPTION
**Pull request by Rich627**

### CODE REVIEW

**Comment by Rich627** [comment]
> Hi @Rich627 請串接 [Get file API](https://aws-educate-tw.postman.co/workspace/API-Development-Workspace~b6915213-64a3-408d-b297-45f472fa0a6b/request/35281217-5f372676-45da-4da5-96c7-7a7a36540395?action=share&source=copy-link&creator=35281217&ctx=documentation)，以透過此 API 來取得某 file 的 detail

Solved, plz review again

File: src/send-email/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# 複製函數代碼
+COPY lambda_function.py /var/task/
+COPY .env /var/task/
+
+# 安裝依賴項
+COPY requirements.txt /var/task/
+RUN pip install -r /var/task/requirements.txt
+
+# 設定運行 Lambda 函數的命令
+CMD ["lambda_function.lambda_handler"]
```
Comment (sh1un) [inline src/send-email/Dockerfile:12]: 這邊安裝 Dependency 的指令 `pip install` 放在後面，而主要執行的程式碼放前面，很可能會影響構建速度，因為你程式碼有改動，後面的 Layer 都要重 Build，建議將安裝 Dependency 的指令放在前面

相關文章:
 - [Docker Layer Cache](https://docs.docker.com/build/cache/)
 - [Docker Best Practices for Python Developers](https://testdriven.io/blog/docker-best-practices/#order-dockerfile-commands-appropriately)

    > Example:
    > 
    > FROM python:3.12.2-slim
    > 
    > WORKDIR /app
    > 
    > COPY sample.py .
    > 
    > COPY requirements.txt .
    > 
    > RUN pip install -r requirements.txt
    > In this Dockerfile, we copied over the application code before installing the requirements. Now, each time we change sample.py, the build will reinstall the packages. This is very inefficient, especially when using a Docker container as a development environment. Therefore, it's crucial to keep the files that frequently change towards the end of the Dockerfile.
    > 
    > You can also help prevent unwanted cache invalidations by using a .dockerignore file to exclude unnecessary files from being added to the Docker build context and the final image. More on this here shortly.
    > 
    > So, in the above Dockerfile, you should move the COPY sample.py . command to the bottom:
    > 
    > FROM python:3.12.2-slim
    > 
    > WORKDIR /app
    > 
    > COPY requirements.txt .
    > 
    > RUN pip install -r requirements.txt
    > 
    > COPY sample.py .

**Review by sh1un — CHANGES_REQUESTED** [review]

File: src/send-email/lambda_function.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        raise
+
+# Function to validate template placeholders against spreadsheet columns
+def validate_template(template_content, columns):
+    try:
+        placeholders = re.findall(r'{{(.*?)}}', template_content) # placeholder format: {{column_name}}
+        missing_columns = [placeholder for placeholder in placeholders if placeholder not in columns]
+        return missing_columns
+    except Exception as e:
+        logger.error(f"Error in excel column, can't find the match placeholder in template: {str(e)}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:66]: Hi @Rich627 
這邊仍有一些地方使用到 f-string 來寫 log，須使用 %s 的方式

File: src/send-email/lambda_function.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                Source=formatted_source_email,
+                Destination={"ToAddresses": [receiver_email]},
+                Message={
+                    "Subject": {"Data": email_title},
+                    "Body": {"Html": {"Data": formatted_content}}
+                }
+            )
+            _ = datetime.datetime.now() + datetime.timedelta(hours=8)
+            formatted_send_time = _.strftime(TIME_FORMAT + "Z")
+            logger.info(f"Email sent to {row.get('Name', 'Unknown')} at {formatted_send_time}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:94]: Hi @Rich627 
同上~

File: src/send-email/lambda_function.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            return formatted_send_time, "SUCCESS"
+        except Exception as e:
+            logger.error(f"Failed to send email to {row.get('Name', 'Unknown')}: {e}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:97]: 同上

File: src/send-email/lambda_function.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            return "FAILED"
+    except Exception as e:
+        logger.error(f"Error in send_email: {str(e)}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:100]: 同上

File: src/send-email/lambda_function.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            'display_name': display_name,
+            'status': status,
+            'recipient_email': recipient_email,
+            'template_file_id': template_file_id,
+            'spreadsheet_file_id': spreadsheet_file_id,
+            'created_at': created_at
+        }   
+        table.put_item(Item=item)
+    except ClientError as e:
+        logger.error(f"Error saving to DynamoDB: {e}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:121]: 同上

File: src/send-email/lambda_function.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    except Exception as e:
+        logger.error(f"Error in save_to_dynamodb: {str(e)}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:123]: 同上

File: src/send-email/lambda_function.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        raise
+
+# Function to handle sending emails and saving results to DynamoDB
+def process_email(ses_client, email_title, template_content, row, display_name, run_id, template_file_id, spreadsheet_id):
+    email = str(row.get("Email", ""))
+    if not re.match(r"[^@]+@[^@]+\.[^@]+", email):
+        logger.warning(f"Invalid email format: {email}. Skipping...") # If email is invalid, skip sending
```
Comment (sh1un) [inline src/send-email/lambda_function.py:130]: 同上

File: src/send-email/lambda_function.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+        # Fetch and read spreadsheet data
+        spreadsheet_info = get_file_info(spreadsheet_id)
+        spreadsheet_s3_key = spreadsheet_info["s3_object_key"]
+        data, columns = read_sheet_data_from_s3(spreadsheet_s3_key)
+
+        # Validate template against spreadsheet columns
+        missing_columns = validate_template(template_content, columns)
+        if missing_columns:
+            logger.error(f"Template validation error: Missing required columns for placeholders: {', '.join(missing_columns)}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:169]: 同上

File: src/send-email/lambda_function.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+            return {"statusCode": 500, "body": json.dumps(response)}
+
+        response = {
+            "status": "SUCCESS",
+            "message": f"All {len(success_recipients)} emails were sent successfully.",
+            "request_id": run_id,
+            "timestamp": datetime.datetime.now().strftime(TIME_FORMAT + "Z"),
+            "sqs_message_id": uuid.uuid4().hex
+        }
+        logger.info(f"Response: {response}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:204]: 同上

File: src/send-email/lambda_function.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+            response = {
+                "status": "FAILED",
+                "message": f"Failed to send {len(failed_recipients)} emails, successfully sent {len(success_recipients)} emails.",
+                "failed_recipients": failed_recipients,
+                "success_recipients": success_recipients,
+                "request_id": run_id,
+                "timestamp": datetime.datetime.now().strftime(TIME_FORMAT + "Z"),
+                "sqs_message_id": uuid.uuid4().hex
+            }
+            logger.info(f"Response: {response}")
```
Comment (sh1un) [inline src/send-email/lambda_function.py:194]: 同上

**Review by sh1un — CHANGES_REQUESTED** [review]

**Review by sh1un — APPROVED** [review]
LGTM!

**Pull request merged**